core/queriec.c

```diff
@@ -0,0 +1,86 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdarg.h>
+
+#include "queriec.h"
+
+void
+queriec_init(struct queriec_obj *queriec, size_t size, size_t offset)
+{
+    queriec->size = size;
+    queriec->status = 0;
+    queriec->offset = offset;
+}
+
+static void
+_queriec_add(struct queriec_obj *queriec, char *query, char *letter, const char key[], size_t key_sz, char value[], size_t value_sz)
+{
+    size_t offset = queriec->offset, i = 0;
+        
+    query[offset++] = letter[0];
+    while (i < key_sz) {
+        query[offset++] = key[i++];
+    }
+    --offset;
+    i = 0;
+    query[offset++] = '=';
+    while (i < key_sz) {
+        query[offset++] = value[i++];
+    }
+}
+
+int queriec_snprintf_add(struct queriec_obj *queriec, char *query, const char key[], size_t keySize, 
+                         char buffer[], size_t bufferLen, const char *format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    size_t supposedSize = (size_t)vsnprintf(buffer, bufferLen, format, args);
+    va_end(args);
+
+    if (queriec->offset + keySize + supposedSize + QUERIEC_ADDITIONAL_LETTERS_SIZE > queriec->size) return QUERIEC_ERROR_NOMEM;
+
+    char *c = NULL;
+    switch (queriec->status) {
+    case 0:
+        c = "?";
+        queriec->status = 1;
+        break;
+    case 1:
+        c = "&";
+        break;
+    default:
+       abort();
+    }
+
+   _queriec_add(queriec, query, c, key, keySize, buffer, supposedSize);
+
+    queriec->offset += keySize + supposedSize + QUERIEC_ADDITIONAL_LETTERS_SIZE;
+
+    return QUERIEC_OK;
+}
+
+int
+queriec_add(struct queriec_obj *queriec, char *query, char key[], size_t keySize, char value[], size_t valueSize)
+{
+    if (queriec->offset + keySize + valueSize + QUERIEC_ADDITIONAL_LETTERS_SIZE > queriec->size) return QUERIEC_ERROR_NOMEM;
+
+    char *c = NULL;
+    switch (queriec->status) {
+    case 0:
+        c = "?";
+        queriec->status = 1;
+        break;
+    case 1:
+        c = "&";
+        break;
+    default:
+       abort();
+    }
+
+   _queriec_add(queriec, query, c, key, keySize, value, valueSize);
+
+    queriec->offset += keySize + valueSize + QUERIEC_ADDITIONAL_LETTERS_SIZE;
```

Comment:
I think it's a good idea to move the last line to inside of `_queriec_add()`, since it seems you'll always need to follow this call with this value update!